apps/passport-client/src/localstorage.ts

```diff
@@ -44,6 +53,20 @@ export async function loadPCDs(self?: User): Promise<PCDCollection> {
     serializedCollection ?? "{}"
   );
 
+  const v3Identity = collection.getPCDsByType(SemaphoreIdentityPCDTypeName);
```

Comment:
This is clever, though a bit awkward to incorporate this kind of app-level state migration in the process of loading from localStorage.  loadInitialState might be a better place to do this.  I wasn't sure until I searched just now but it seems that's the only place where loadPCDs is actually called now.  If there were more calls then there might be legit worry about whether or not to migrate in all of them.

If there was a single identity PCD containing V3 and V4, then this step would happen inside of its deserialize() function, which feels well encapsulated, and ensures it happens in all cases.

Reply:
> Having a deserialize() function do something other than just deserializing the data would be quite surprising to me if I were the caller. It would mean that if I deserialized something and serialized it again, the output would not be the same as the original serialized input which, while not a hard rule, seems like a sensible thing to expect, except in cases where serialization formats themselves change.

Not the response I expected, so let's definitely discuss this.  I think what's implied by my proposal here is that I'm framing this migration as a format version upgrade, which is something I think PCD package deserialization is responsible for handling.  SemaphoreIdentityPCD represent's a user's identity.  In it's current version (call it V3 for convenience) that's represented by a V3 trapdoor+nullifier.  In the new version (V4) it's represented by two things: V3 secrets, and also a V4 private key.  The new thing can be derived from the old thing, so that can happen silently as an upgrade at deserialization time.  There could be an explicit version number in the data to drive this, but in this case the presence or absence of a new field is sufficient to detect the version and do the upgrade.  New code only ever sees the upgraded version.  This approach is attractive to me because it naturally handles all cases.  At the moment that the code upgrades to pick up the changes in this PR, there can't be any data in memory, so all data which enters memory has to go through deserialization to get there.  The upgrade happens regardless of where a PCD comes from: local storage, E2EE storage, or even a URL request to Add a PCD coming from another app with old code.

It would be possible and equivalent to do the upgrade when building the claims after deserialization (not really a distinct step right now), or lazily any time the V4 secret is requested (awkward because PCDs don't generally have accessor functions, just fields).  We could also approach this such that the V4 secret is never saved at all, but just generated on-demand, since it's derived data (at least for private Identity).  I wouldn't go quite that far since the derivation has a cost, and our intent is to eventually get rid of the V3 secret entirely.

Reply:
For a change which is hard (though maybe not inpossible) to reverse later, I'd argue to optimize for reduced complexity and fragility in future, even at the cost of some extra effort now.  I worry about how many different places in the code handle the migration now leading to one of them breaking subtly in future.  E.g when someone adds a new type of login flow and don't handle migration correctly, which will break only for the few users who haven't migrated yet.  I also worry about whether we'll lose the ability to treat the 2 PCDs as one identity as easily in future, though the function to derive V4 from V3 does let us find them and link them together if we really need to.

Those are theoretical worries, though, and I'm not going to push it if you're still hesitant.  I'm sure I'm irrationallly attached to the idea because it's my idea.  I made a suggestion elsewhere about ways to simplify and centralize some of the migration code (outside of localStorage), which is what I'd suggest you focus on if you want an alternative.